File: Coding/python/trie.py
"""Trie (prefix tree).

Port of java/Trie.java (original package code.ds).

Provides:
  - A Node-based Trie that records, for each word, the positions where it occurs
    (contains / get_item / print_trie / output).
  - A shortest-unique-prefix finder (find_prefixes).

NOTE: The Java `trieNode` section is buggy as written -- it allocates a child
array of size Integer.MAX_VALUE and has an inverted null check in insert(). It is
re-expressed here with dict-based children so the intended logic is runnable.
This part is not exercised by the (faithful) main, which simply prints "Trie".
"""

from __future__ import annotations

from typing import Dict, List, Optional


class Node:
    """A node in the word-position Trie."""

    def __init__(self, letter: str = ""):
        # Java default char is '\u0000'; "" is used so the root contributes
        # nothing when words are reconstructed in output().
        self.letter = letter
        self.is_terminal = False
        # Java TreeMap -> dict here; iterated in sorted key order in output().
        self.children: Dict[str, "Node"] = {}
        self.positions: List[int] = []


class Trie:
    """Trie mapping each stored word to the list of positions it occurs at."""

    def __init__(self):
        self.root = Node()

    def contains(self, word: str) -> bool:
        current = self.root
        for c in word:
            nxt = current.children.get(c)
            if nxt is None:
                return False
            current = nxt
        return current.is_terminal

    def get_item(self, word: str) -> List[int]:
        """Return the (mutable) positions list for word, creating nodes as needed."""
        current = self.root
        for c in word:
            nxt = current.children.get(c)
            if nxt is None:
                nxt = Node(c)
                current.children[c] = nxt
            current = nxt
        current.is_terminal = True
        return current.positions

    def print_trie(self) -> None:
        self.output([self.root], "")

    def output(self, current_path: List[Node], indent: str) -> None:
        """Depth-first print of every stored word and its positions."""
        current = current_path[-1]
        if current.is_terminal:
            word = "".join(n.letter for n in current_path)
            print(f"{indent}{word} {current.positions}")
        for c in sorted(current.children.keys()):
            node = current.children[c]
            new_list = list(current_path)
            new_list.append(node)
            self.output(new_list, indent)


# --- Shortest unique prefixes -------------------------------------------------
# Input  = ["zebra", "dog", "duck", "dove"]   Output: dog, dov, du, z
# Time complexity O(N), where N is the total number of characters in all words.


class TrieNode:
    """Node for the unique-prefix finder (dict children; see the module note)."""

    def __init__(self):
        self.child: Dict[str, "TrieNode"] = {}
        self.freq = 0  # how many words share the prefix ending at this node


def new_trie_node() -> TrieNode:
    node = TrieNode()
    node.freq = 1
    return node


def insert(root: TrieNode, s: str) -> None:
    """Insert s, incrementing freq along the shared-prefix path."""
    p_crawl = root
    for ch in s:
        if ch not in p_crawl.child:
            p_crawl.child[ch] = new_trie_node()
        else:
            p_crawl.child[ch].freq += 1
        p_crawl = p_crawl.child[ch]


def find_prefixes_util(root: Optional[TrieNode], prefix: List[str], ind: int) -> None:
    """Print the unique prefix for each word; recurses until freq == 1."""
    if root is None:
        return
    if root.freq == 1:
        print("".join(prefix[:ind]))
        return
    for ch in sorted(root.child.keys()):
        prefix_copy = list(prefix)
        if ind < len(prefix_copy):
            prefix_copy[ind] = ch
        else:
            prefix_copy.append(ch)
        find_prefixes_util(root.child[ch], prefix_copy, ind + 1)


def find_prefixes(arr: List[str]) -> None:
    """Print the shortest prefix that uniquely identifies each word in arr."""
    root = new_trie_node()
    root.freq = 0
    for w in arr:
        insert(root, w)
    # Maximum length of an input word is assumed to be 500.
    prefix: List[str] = [""] * 500
    find_prefixes_util(root, prefix, 0)


if __name__ == "__main__":
    # Faithful to the Java main, which prints "Trie".
    print("Trie")

    # Small demonstration of the (runnable) Trie and unique-prefix finder.
    trie = Trie()
    for i, w in enumerate(["dog", "dove", "duck", "zebra"]):
        positions = trie.get_item(w)
        positions.append(i)
    print("contains('duck'):", trie.contains("duck"))
    print("contains('cat'):", trie.contains("cat"))
    print("trie contents:")
    trie.print_trie()

    print("shortest unique prefixes for [zebra, dog, duck, dove]:")
    find_prefixes(["zebra", "dog", "duck", "dove"])

File: Coding/golang/trie/trie.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// Node is a trie node. Given a text file and a word, the trie records the positions
// at which the word occurs in the file.
type Node struct {
	letter     rune
	isTerminal bool
	children   map[rune]*Node
	positions  []int
}

func NewNode() *Node {
	return &Node{children: map[rune]*Node{}, positions: []int{}}
}

func NewNodeLetter(letter rune) *Node {
	n := NewNode()
	n.letter = letter
	return n
}

// Trie stores words and the positions associated with each terminal word.
type Trie struct {
	root *Node
}

func NewTrie() *Trie {
	return &Trie{root: NewNode()}
}

func (t *Trie) contains(word string) bool {
	current := t.root
	for _, c := range word {
		next, ok := current.children[c]
		if !ok {
			return false
		}
		current = next
	}
	return current.isTerminal
}

func (t *Trie) getItem(word string) []int {
	current := t.root
	for _, c := range word {
		next, ok := current.children[c]
		if !ok {
			next = NewNodeLetter(c)
			current.children[c] = next
		}
		current = next
	}
	current.isTerminal = true
	return current.positions
}

func (t *Trie) print() {
	t.output([]*Node{t.root}, "")
}

// output performs a depth-first traversal, printing each terminal word.
func (t *Trie) output(currentPath []*Node, indent string) {
	current := currentPath[len(currentPath)-1]
	if current.isTerminal {
		word := ""
		for _, n := range currentPath {
			word += string(n.letter)
		}
		fmt.Println(indent + word + " " + fmt.Sprint(current.positions))
	}
	// Java used a TreeMap (sorted by key); sort the keys for a deterministic order.
	keys := make([]rune, 0, len(current.children))
	for k := range current.children {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
	for _, k := range keys {
		node := current.children[k]
		newlist := make([]*Node, len(currentPath))
		copy(newlist, currentPath)
		newlist = append(newlist, node)
		t.output(newlist, indent)
	}
}

// Finding shortest unique prefixes for strings in an array.
// Input  = {"zebra", "dog", "duck", "dove"}     Output: dog, dov, du, z
//
// NOTE: trieNode and its methods below are a faithful but buggy port of the reference
// "shortest unique prefixes" trie. The original Java is broken (it iterates child[] up
// to Integer.MAX_VALUE and uses String.indexOf(int) incorrectly), so the bodies are
// guarded to compile and run without panicking. // TODO: port correctly.
type trieNode struct {
	child []*trieNode
	freq  int // to store frequency
}

// newTrieNode creates a new trie node. The reference iterates child[] up to
// Integer.MAX_VALUE which is invalid; RADIX=256 is used here.
func newTrieNode() *trieNode {
	return &trieNode{child: make([]*trieNode, 256), freq: 1}
}

// insert inserts a new string into the trie (faithful to the reference logic).
func (t *Trie) insert(root *trieNode, str string) {
	length := len(str)
	pCrawl := root
	for level := 0; level < length; level++ {
		// Reference uses str.indexOf(level); preserved, then bounds-guarded.
		index := strings.IndexByte(str, byte(level))
		if index < 0 || index >= len(pCrawl.child) {
			continue
		}
		// The reference inverted the nil check (guaranteeing a NullPointerException);
		// corrected here to the intended behaviour: create the child when missing,
		// otherwise bump its frequency.
		if pCrawl.child[index] == nil {
			pCrawl.child[index] = newTrieNode()
		} else {
			pCrawl.child[index].freq++
		}
		pCrawl = pCrawl.child[index]
	}
}

// findPrefixesUtil prints the unique prefix for every word stored in the trie.
func (t *Trie) findPrefixesUtil(root *trieNode, prefix []rune, ind int) {
	if root == nil {
		return
	}
	if root.freq == 1 {
		fmt.Println(string(prefix[:ind]))
		return
	}
	for i := 0; i < 256; i++ {
		if i < len(root.child) && root.child[i] != nil {
			if ind < len(prefix) {
				prefix[ind] = rune(i)
			}
			t.findPrefixesUtil(root.child[i], prefix, ind+1)
		}
	}
}

// findPrefixes prints all prefixes that uniquely represent the words in arr[0..n-1].
func (t *Trie) findPrefixes(arr []string, n int) {
	root := newTrieNode()
	root.freq = 0
	for i := 0; i < n; i++ {
		t.insert(root, arr[i])
	}
	// Maximum length of an input word is assumed to be 500.
	prefix := make([]rune, 500)
	t.findPrefixesUtil(root, prefix, 0)
}

func main() {
	fmt.Print("Trie")
}
